[Gordon Yit]iP

Gordon25 opened the pull request:
# Duke 2.0
>"Your mind is like Software. Program it"
> -Peter Thiel

## Duke2.0 is like your mind, *programmed* to remember things *for* you. It is :
*Fast and Efficient
*easy to use
*texted-based

## All you need to do is,
1. Download it from [here](https://github.com/nus-cs2103-AY2122S1/ip/pull/228)
2. double click it.
3. add your tasks to it
4. Let it handle everything else 😁 

## And its **_FREE_**

Want to know more about Duke2.0? Check out our **main** attraction:
```  java  
    public static void main(String[] args) {
        new Duke("data", "duke.txt").run();
    }
```

Commit:
Add Gradle support

Gordon25 committed:
Added ability to greet, echo and exit

Gordon25 committed:
Added javadocs and indented output of the chatbot

Gordon25 committed:
Added ability to store text and list the text stored in order

Gordon25 committed:
Added task class

Gordon25 committed:
Added ability for Duke to support task class, mark task as done

Gordon25 committed:
Added more javadoc to task

Gordon25 committed:
Added Deadline, Event and Todo classes.

Gordon25 committed:
Add ability to support tracking of Deadline, Event and Todo tasks

Gordon25 committed:
Added a private method formatInput to Duke class

Gordon25 committed:
Added test inputs and expected outputs for Uitesting

Gordon25 committed:
Removed some escape characters in Duke class

Gordon25 committed:
Added DukeException class

Gordon25 committed:
Added ability to handle exceptions

Gordon25 committed:
Added testing for exception handling

Gordon25 committed:
Changed exception message for StringIndexOutOfBoundsException in DukeException class

Gordon25 committed:
Added ability to delete tasks

Gordon25 committed:
Change exception message for StringIndexOutOfBoundsException Exception

Gordon25 committed:
Change input commands

Gordon25 committed:
Update expected outputs

Gordon25 committed:
Update javadocs

Gordon25 committed:
Add enums and update javadocs

Gordon25 committed:
Update enum and switch statements

Gordon25 committed:
Add task marker method

Gordon25 committed:
Change string format of task classes

Gordon25 committed:
Add storage file and DataFile class

Gordon25 committed:
Add ability to save tasks into hard disk

Gordon25 committed:
Add DateDue class

Gordon25 committed:
Add ability to interpret dates.

Gordon25 committed:
Merge branch 'branch-Level-7'

Gordon25 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/DukeException.java
#	src/main/java/Event.java
#	text-ui-test/EXPECTED.TXT
#	text-ui-test/input.txt

Gordon25 committed:
Add Command (Exit, Add, Filter, List, Delete and MarkDone), Date, Ui, Parser, Storage, TaskList classes.
Remove DataFile and DateDue classes.

Gordon25 committed:
Reorganise duke and supporting classes into Ui, Parser, Command, Task, Storage and Tasks List classes.

Gordon25 committed:
Reorganise classes into duke.main, duke.command, duke.task packages.

Gordon25 committed:
Add Test classes for methods in DukeException and TaskList classes.

Gordon25 committed:
Change directory to ensure JAR file executes properly

Gordon25 committed:
Add gitignore file

Gordon25 committed:
Add javadocs

Gordon25 committed:
Add javadocs to test classes

Gordon25 committed:
Revert "Add javadocs to test classes"

This reverts commit 43aec2044d511bbc1755260a2b67a0d1f3575079.

Gordon25 committed:
Add javadocs for test classes

Gordon25 committed:
Follow coding standard

Gordon25 committed:
Add FindCommand class

Gordon25 committed:
Add ability to find task using search phrases

Gordon25 committed:
Merge branch 'branch-A-JavaDoc'

Gordon25 committed:
Merge branch 'branch-A-CodingStandard'

Gordon25 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/main/TaskList.java
#	src/test/java/duke/main/TaskListTest.java

eltonyeh commented on src/main/java/IP/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,61 @@
+     * @param tasks lists of tasks
+     * @param ui the user interface.
+     * @param storage the storage file.
+     * @throws DukeException
```
Will it be better if you specify when will the exception be thrown?
```suggestion
     * @throws DukeException If ...
```

eltonyeh commented on src/main/java/IP/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,61 @@
+import IP.duke.task.Deadline;
+import IP.duke.task.Event;
+import IP.duke.task.Task;
+import IP.duke.task.Todo;
```
Shouldn't the name representing packages be in all lower case?

eltonyeh commented on src/main/java/IP/duke/command/FilterCommand.java:
```diff
@@ -0,0 +1,40 @@
+    
+    /**
+     * Executes a command to filter out tasks falling on the specified date.
+     * @param tasks lists of tasks
```
Should there be a blank line between these two lines?

eltonyeh commented on src/main/java/IP/duke/main/Date.java:
```diff
@@ -0,0 +1,72 @@
+     */
+    public Date(String dateString) throws DateTimeParseException {
+        date = dateString.split("/");
+            localDate = LocalDate.parse(String.format("%s-%s-%s", "2021", date[1], date[0]));
```
Is the indentation here wrong?

eltonyeh commented on src/main/java/IP/duke/main/Date.java:
```diff
@@ -0,0 +1,72 @@
+    /**
+     * Retrieves the local date object.
+     * 
+     * return the local date of the date due.
```
Should there be a `@` symbol in the beginning?
```suggestion
     * @return the local date of the date due.
```

eltonyeh commented on src/main/java/IP/duke/main/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
+        while(!isExit) {
```
Should there be a space between `while` and `(`?
```suggestion
        while (!isExit) {
```

eltonyeh commented on src/main/java/IP/duke/main/DukeException.java:
```diff
@@ -0,0 +1,47 @@
+ * @author Gordon Yit
+ * @version CS2103T, Semester 2.
+ */
+public class DukeException extends Exception{
```
Should there be a space between `Exception` and `{`?
```suggestion
public class DukeException extends Exception {
```

eltonyeh commented on src/main/java/IP/duke/main/Parser.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class Parser {
+    
+    public enum Commands { done, list, delete, filter, others};
```
Should the brackets be in Egyptian style and without the semicolon?
```suggestion
    public enum Commands { 
        done, list, delete, filter, others
    }
```

eltonyeh commented on src/main/java/IP/duke/main/Storage.java:
```diff
@@ -0,0 +1,116 @@
+     * @return an arraylist of tasks.
+     * @throws IOException
+     */
+    public  ArrayList<Task> load() throws DukeException {
```
```suggestion
    public ArrayList<Task> load() throws DukeException {
```

eltonyeh commented on src/main/java/IP/duke/main/Storage.java:
```diff
@@ -0,0 +1,116 @@
+        ArrayList<Task> tasks = new ArrayList<>();
+        try {
+            File fileDirectory = new File(filePath);
+            if(!fileDirectory.exists()) {
```
Should there be a space between `if` and `(`? I also spotted this in different places.
```suggestion
            if (!fileDirectory.exists()) {
```

eltonyeh commented on src/main/java/IP/duke/main/TaskList.java:
```diff
@@ -0,0 +1,105 @@
+    public ArrayList<Task> findTasksMatchingDate(Date date) {
+        String dateString = date.toString();
+        ArrayList<Task> tasksMatchingDate = new ArrayList<>();
+        for (Task t: tasks) {
```
```suggestion
        for (Task t : tasks) {
```

eltonyeh commented on src/main/java/IP/duke/task/Deadline.java:
```diff
@@ -0,0 +1,88 @@
+ * @version CS2103T, Semester 1
+ */
+public class Deadline extends Task {
+    private final String TASK_MARKER = "D";
```
Not a coding standard violation, but should the constant be declared `static`?

eltonyeh commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,61 @@
+            }
+            tasks.add(task);
+            ui.showTaskAdded(task, tasks.getTasks());
+        } catch (StringIndexOutOfBoundsException e) {
+            throw new DukeException(e);
+        } catch (DateTimeParseException e) {
+            throw new DukeException(e);
+        }
```
Not a coding standard violation, but is it better to merge the two catch blocks?
```suggestion
        } catch (StringIndexOutOfBoundsException | DateTimeParseException e) {
            throw new DukeException(e);
        }
```

eltonyeh commented on src/main/java/duke/main/Date.java:
```diff
@@ -0,0 +1,72 @@
+     * converts a given date string into a date object.
+     * 
+     * @param dateString in form of DD MMM YYYY.
+     * @return a date object corresponding to the date string.
```
Maybe there should be a `@throw` line after it?

eltonyeh commented on src/main/java/duke/main/Date.java:
```diff
@@ -0,0 +1,72 @@
+     */
+    public static Date convertDateStringToDate(String dateString) throws ParseException {
+        SimpleDateFormat stringDateFormat = new SimpleDateFormat("MMM D");
+        java.util.Date s = stringDateFormat.parse(dateString);
```
Should the naming of `s` be more meaningful?

eltonyeh commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    }
+    
+    /**
+     * Runs the Duke.Duke chat bot.
+     * @param args Unused.
```
```suggestion
     * Runs the Duke.Duke chat bot.
     *
     * @param args Unused.
```

eltonyeh commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,57 @@
+ */
+public class Parser {
+    
+    public enum Commands { done, list, delete, filter, others};
```
```suggestion
    public enum Commands { 
        done, list, delete, filter, others
    }
```

eltonyeh commented on src/test/java/IP/duke/main/DukeExceptionTest.java:
```diff
@@ -0,0 +1,32 @@
+    public void getMessage_StringOutOfBoundsException_CommandIncompleteMessage(){
+        StringIndexOutOfBoundsException e = new StringIndexOutOfBoundsException();
+        DukeException dukeException = new DukeException(e);
+        assertEquals(dukeException.getMessage(), "☹ The task description or command is incomplete.:$\n");
```
Not a standard violation, but shouldn't the expected result be the first parameter?
```suggestion
        assertEquals("☹ The task description or command is incomplete.:$\n", dukeException.getMessage());
```

eltonyeh reviewed (commented):
Almost LGTM 👍 ! There are some spacing/empty line issues to be fixed, but most of them are minor issues.

Gordon25 committed:
Change error message in DukeException class.

Gordon25 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

wanninglim commented on src/main/java/IP/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,61 @@
+            }
+            tasks.add(task);
+            ui.showTaskAdded(task, tasks.getTasks());
+        } catch (StringIndexOutOfBoundsException e) {
```
I think you can catch both exceptions in a single catch block :-)

```suggestion
        } catch (StringIndexOutOfBoundsException | DateTimeParseException e) {
```

wanninglim commented on src/main/java/IP/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,50 @@
+     * @param tasks lists of tasks
+     * @param ui the user interface.
+     * @param storage the storage file.
+     * @throws DukeException
```
Same as what @eltonyeh mentioned, I think it might be clearer if you add an if... line here!

```suggestion
     * @throws DukeException if...
```

wanninglim commented on src/main/java/IP/duke/main/Parser.java:
```diff
@@ -0,0 +1,54 @@
+            default:
+                    return new AddCommand(command);
+            }
+        } catch (ArrayIndexOutOfBoundsException e) {
```
Same here, I think you can possibly combine these exceptions into a single catch block if it has the same body! :D

wanninglim commented on src/main/java/IP/duke/main/Storage.java:
```diff
@@ -0,0 +1,116 @@
+                        t = new Deadline(task, time);
+                    } else if (line.contains("E |")) {
+                        String taskDescriptionAndTime = line.substring(startingIndex);
+                        int startOfTimeIndex = taskDescriptionAndTime.indexOf(divider);
```
I'm not very sure which is more efficient, but it seems like the code would look shorter if you can extract the 4 common lines for "D |" and "E |" into another if statement block.

also, I think `String taskDescriptionAndTime = line.substring(startingIndex);` could be extracted out of this 3 blocks since it is common to all conditionals! 👯

wanninglim commented on src/main/java/IP/duke/main/Ui.java:
```diff
@@ -0,0 +1,160 @@
+    private String buffer1 = c.repeat((width - firstLine.length()) / 2);
+    private String buffer2 = c.repeat((width - secondLine.length()) / 2);
+    private String welcomeMessage = buffer1 + firstLine + buffer1 + "\n" + buffer2 + secondLine + buffer2 + "\n";
+    private final String GREETING = "Hello from\n" + LOGO + welcomeMessage;
```
Can possibly group constants at the start of the file instead of after normal variables. I think checkstyle checks for this so once you add checkstyle it should detect!

wanninglim commented on src/main/java/IP/duke/task/Event.java:
```diff
@@ -0,0 +1,88 @@
+ * @version CS2103T, Semester 2
+ */
+public class Event extends Task {
+    private final String TASK_MARKER = "E";
```
Same as what @eltonyeh mentioned, if your var is capitalised (you're declaring it as a constant) I think it should be declared as `static final`!

wanninglim reviewed (commented):
Hello! Generally looks good to me but you can take note of some places where your code can be simplified further. 👍 Also for clarity and neatness, for longer code (like in Storage.java), you can also consider doing line breaks to group up related code if necessary!

Gordon25 committed:
Delete duplicate files

Gordon25 committed:
Add Gradle

Gordon25 committed:
Delete duplicate files

Gordon25 committed:
Change exception handling in dukeException, change build.gradle file to enable gradle to run JUnit Tests

Gordon25 committed:
Add FindCommad class, change logic of Command, TaskList and Ui classes.

Gordon25 committed:
Merge branch 'branch-A-Gradle'

# Conflicts:
#	build.gradle
#	src/main/java/IP/duke/command/AddCommand.java
#	src/main/java/IP/duke/command/DeleteCommand.java
#	src/main/java/IP/duke/command/FilterCommand.java
#	src/main/java/IP/duke/main/Storage.java
#	src/main/java/IP/duke/main/TaskList.java
#	src/main/java/IP/duke/main/Ui.java

Gordon25 committed:
Add javaDocs

Gordon25 committed:
Use gradle to run checkstyle checks

Gordon25 committed:
Rename IP.duke package to duke package, start on javafx

Gordon25 committed:
Change rename package from IP.duke to duke

Gordon25 committed:
Add skeletal GUI

Gordon25 committed:
Add Varags

Gordon25 committed:
Change stub to actual duke logic

Gordon25 committed:
Change photo border and text positioning.

Gordon25 committed:
Merge branch 'Branch-A-CheckStyle'

Gordon25 committed:
Merge branch 'Branch-A-Varags'

# Conflicts:
#	build.gradle
#	src/main/java/IP/duke/command/FilterCommand.java
#	src/main/java/duke/main/Date.java
#	src/main/java/duke/main/Ui.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/test/java/IP/duke/main/TaskListTest.java

Gordon25 committed:
Merge branch 'branch-A-Level-10'

# Conflicts:
#	build.gradle
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/FilterCommand.java
#	src/main/java/duke/command/FindCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkDoneCommand.java
#	src/main/java/duke/main/Parser.java
#	src/main/java/duke/main/Storage.java
#	src/main/java/duke/main/TaskList.java
#	src/main/java/duke/main/Ui.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java
#	src/test/java/duke/main/DukeExceptionTest.java

Gordon25 committed:
Follow checkstyle

Gordon25 committed:
Create shadow jar file

Gordon25 committed:
Add assertion statements

Gordon25 committed:
duke.Command : refactor code

Methods in classes of duke.Command package have more than one level
of abstraction.

A single level of abstraction in one method improves readability.

Let's
* update the methods in classes of duke.Command package to have only
one level of abstraction.

Gordon25 committed:
duke package: refactor code

Classes have long methods and methods have more than one level of
abstraction. Some methods have arrow head code style.

Let's divide long methods into smaller methods, each with single
level of abstraction. Nested if else statements replaced with guard
clauses, long if else statements replaced with switch statements.

Short methods, single level of abstraction and long happy path
greatly improves readability.

Gordon25 committed:
duke package : Add assert statements

There is no assertion statements to check if the code behaves
according to my assumptions.

Assert statements help to validate assumptions made about the
program and flag out any assumptions that do not hold.

Let's add assert statements to check if variables are correctly
assigned before they are operated on and the variables take on
reasonable values.

Gordon25 committed:
Add more assert statements

Gordon25 committed:
Merge pull request #1 from Gordon25/branch-A-Assertions

Merge with branch-A-Assertions

Gordon25 committed:
Refactor gui classes under duke.gui, rename variables to improve
readability.

Gordon25 committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/FilterCommand.java
#	src/main/java/duke/command/FindCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkDoneCommand.java
#	src/main/java/duke/gui/DukeApplication.java
#	src/main/java/duke/main/Date.java
#	src/main/java/duke/main/Parser.java
#	src/main/java/duke/main/Storage.java
#	src/main/java/duke/main/Ui.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/task/Todo.java
#	src/test/java/IP/duke/main/TaskListTest.java

Gordon25 committed:
Add ability to tag a task when adding the task

Gordon25 committed:
Merge pull request #4 from Gordon25/branch-A-BCD-Extension

Merge branch-A-BCD-Extension branch to master

Gordon25 committed:
Add TagCommand class

Gordon25 committed:
Add ability to tag tasks

Gordon25 committed:
Merge pull request #5 from Gordon25/branch-A-BCD-Extension

Merge Branch a bcd extension to master

Gordon25 committed:
Create workflows

Gordon25 committed:
Delete .github directory

Gordon25 committed:
Merge pull request #6 from Gordon25/Gordon25-patch-1

Delete .github directory

Gordon25 committed:
Create gradle.yml

Gordon25 committed:
Merge pull request #7 from Gordon25/Gordon25-patch-1-1

Create gradle.yml

Gordon25 committed:
Updated gradle.yml file

Gordon25 committed:
Added more changes to comply with checkstyle

Gordon25 committed:
Add newline to file

Gordon25 committed:
Update test methods for DukeException

Gordon25 committed:
Fix issue with "bye" command on gui

Gordon25 committed:
Refactor Storage#createNewTask method

Gordon25 committed:
Refactor DukeApplication code into Duke class
Fix issue with welcome message not displayed when initializing chatbot

Gordon25 committed:
Fix issue of stored tasks not loaded upon list command after starting
chatbot

Gordon25 committed:
Delete Date class

Date class overlaps in logic with TaskDate class

Gordon25 committed:
Fix issue tasks with no tags and handling improper date format

Gordon25 committed:
Enhance GUI

Make GUI asymmetric
Trimmed out profile photo background
Highlight error messages

Gordon25 committed:
Do more refactoring

Gordon25 committed:
Add Product website

Gordon25 committed:
Update README to include reuse sources

Gordon25 committed:
Replace acknowledgement segment in correct README file

Gordon25 committed:
Complete Userguide with user functions
Add acknowledgement segment to show source of reused code

Gordon25 committed:
Fix issue with application not terminating after exit command

Gordon25 committed:
Fix storage to handle cases when tasks are not tagged

Gordon25 committed:
Reformat User guide

Gordon25 committed:
Move product website to docs folder

Gordon25 committed:
Add Profile pictures

Gordon25 committed:
README.md : Fix formatting issues

Gordon25 committed:
Update README.md

Gordon25 committed:
Update README.md

Gordon25 committed:
Update README.md

Gordon25 committed:
Update README.md

Gordon25 committed:
Update README.md

Gordon25 committed:
Update README.md

Gordon25 committed:
Update README.md

Gordon25 committed:
Add ability to save task even if user clicks on close window button

Gordon25 committed:
Merge branch 'master' of https://github.com/Gordon25/ip
